src/goto-programs/skolemization.cpp

```diff
@@ -0,0 +1,69 @@
+/// \file skolemization.h
+/// Rename variables in existentially quantified statements.
+/// Author: Diffblue Ltd.
+
+#include "skolemization.h"
+
+#include "goto_model.h"
+
+#include <ansi-c/expr2c.h>
+
+#include <util/expr_util.h>
+#include <util/fresh_symbol.h>
+#include <util/irep.h> // for id2string
+#include <util/mathematical_expr.h>
+
+void rename_variables_in_existential_quantifiers(goto_modelt &goto_model)
+{
+  for(auto &function : goto_model.goto_functions.function_map)
+  {
+    for(auto &instruction : function.second.body.instructions)
+    {
+      if(instruction.has_condition())
+      {
+        auto &cond = instruction.condition_nonconst();
+        recurse_on_condition(cond, goto_model);
+      }
+    }
+  }
+}
+
+void recurse_on_condition(exprt &cond, goto_modelt &model)
+{
+  // recurse on the condition's subexpressions first - that allows us
+  // to correctly identify the nodes we care about in a compound
+  // expression, for instance something like `1 == 1 && exists(i.i>=0 && i>1)`
+  for(exprt &subexpr : cond.operands())
+  {
+    recurse_on_condition(subexpr, model);
+  }
+
+  // if the expression we're in doesn't have any more operands,
+  // it means it's a leaf node, so we go ahead and process it.
```

Comment:
:no_entry: I find this comment really misleading. The following statement will be reached whether or not this is a leaf node. In fact the nodes of interest are not leaf nodes, as they are exists quantifiers and these quantifier expressions have child nodes. I think the phrase you are after is "depth first search".

Reply:
This comment is now gone.